### PR TITLE
Look-up value for JavaHome through Windows registry to load jvm.dll

### PR DESCRIPTION
**Pull request by AndyGreenwell**
This commit makes use of the nascent [WinReg.jl](https://github.com/simonbyrne/WinReg.jl) functionality developed by @simonbyrne (not yet a registered package) to lookup the location of the JavaHome registry entry and then load jvm.dll based on the determined location of that path, instead of requiring that the user must have set a JAVA_HOME environment variable.  This PR should likely not be merged until the WinReg.jl package is available through METADATA, but submitting now to allow for others to review the changes.

This commit also loads the copy of msvcr100.dll that is packaged with the JDK on Windows.  At present, testing has only been performed on Windows systems with JDK 1.8 installed.  Consequently, older and newer versions of jvm.dll might include a different version of the msvcr dll in their packaging that would need to be loaded prior to loading jvm.dll.  A more robust manner of detecting which version of msvc is packaged with the JDK that should be loaded prior to loading jvm.dll is likely needed.  

Testing has been performed on the following combinations of Julia/Java/Windows.  64-bit Julia/Java/Windows.  32-bit Julia/Java/Windows.  32-bit Julia/Java with 64-bit Windows.

### CODE REVIEW

File: src/jvm.jl [hunk 1]
```diff
@@ -16,17 +16,47 @@ const JNI_ENOMEM       = convert(Cint, -4)              #/* not enough memory */
...
 
+function javahome_winreg()
+    processor_architecture = ENV["PROCESSOR_ARCHITECTURE"]
+    keypath = utf16("")
+    javahome = ""
+    cv = "CurrentVersion"
+    jh = "JavaHome"
+    global const msvclib = "msvcr100.dll"
+    
+    if processor_architecture == "AMD64"
```
Comment (simonbyrne) [inline src/jvm.jl]: I thought that we found at today that the env `PROCESSOR_ARCHITECTURE` is dependant on the word size of the application?

File: src/jvm.jl [hunk 2]
```diff
@@ -16,17 +16,47 @@ const JNI_ENOMEM       = convert(Cint, -4)              #/* not enough memory */
...
+    javahome = ""
+    cv = "CurrentVersion"
+    jh = "JavaHome"
+    global const msvclib = "msvcr100.dll"
+    
+    if processor_architecture == "AMD64" 
+        if WORD_SIZE == 64
+            keypath = utf16("SOFTWARE\\JavaSoft\\Java\ Development\ Kit")
+        elseif WORD_SIZE == 32
+            keypath = utf16("SOFTWARE\\WoW6432Node\\JavaSoft\\Java\ Development\ Kit")
```
Comment (simonbyrne) [inline src/jvm.jl]: From some quick googling, it seems as though we shouldn't need to do this e.g. see https://msdn.microsoft.com/en-us/library/windows/desktop/aa384232(v=vs.85).aspx

File: src/jvm.jl [hunk 3]
```diff
@@ -16,17 +16,47 @@ const JNI_ENOMEM       = convert(Cint, -4)              #/* not enough memory */
...
-    
+    msvcpaths = Any[]
+      
+    @windows_only push!(javahomes, javahome_winreg())
+
     if haskey(ENV,"JAVA_HOME")
         push!(javahomes,ENV["JAVA_HOME"])
     end
-    if isfile("/usr/libexec/java_home")
+    if isexecutable("/usr/libexec/java_home")
```
Comment (tkelman) [inline src/jvm.jl]: `isexecutable` is deprecated

File: src/jvm.jl [hunk 4]
```diff
@@ -38,16 +68,28 @@ function findjvm()
...
         @linux_only if WORD_SIZE==32; push!(libpaths, joinpath(n, "jre", "lib", "i386", "server")); end
         push!(libpaths, joinpath(n, "jre", "lib", "server"))
     end
     
     ext = "."*@windows? "dll":@osx? "dylib":"so"
+    
+    try 
+        for n in msvcpaths
+            msvcpath = joinpath(n, msvclib)
+            if isreadable(msvcpath)
```
Comment (tkelman) [inline src/jvm.jl]: as is `isreadable`

File: src/jvm.jl (hunk 3 above)
Comment (AndyGreenwell) [inline src/jvm.jl]: Had my local environment working from the latest tagged release of JavaCall (0.3.4), not master.  Will fix.

**Comment by aviks** [comment]
Should I merge this, or will WinReg.jl be in METADATA soon?

**Comment by tkelman** [comment]
Calling Pkg operations from src is not appropriate. The code should either be registered or copied.

**Comment by simonbyrne** [comment]
see https://github.com/JuliaLang/METADATA.jl/pull/5029

Also, you shouldn't need to use `utf16`: `querykey` will do the conversion automatically.

**Comment by AndyGreenwell** [comment]
Tests for Julia version 0.3 are failing since WinReg.jl currently only supports Julia 0.4 and later.  

@simonbyrne @aviks Should WinReg.jl be altered to support 0.3, or should JavaCall's supported version be bumped to 0.4, or should this PR be altered to be special cased for Julia versions prior to 0.4 and versions later than 0.4?

**Comment by simonbyrne** [comment]
It should be possible to support Julia 0.3, but it's not as straightforward as changing version and adding Compat.jl. 

@aviks Are you planning on continuing support JavaCall for 0.3?

File: src/jvm.jl [hunk 5]
```diff
@@ -16,13 +16,24 @@ const JNI_ENOMEM       = convert(Cint, -4)              #/* not enough memory */
 const JNI_EEXIST       = convert(Cint, -5)              #/* VM already created */
 const JNI_EINVAL       = convert(Cint, -6)              #/* invalid arguments */
 
+function javahome_winreg()
+    global const msvclib = "msvcr100.dll"
+    keypath = "SOFTWARE\\JavaSoft\\Java\ Development\ Kit"
```
Comment (simonbyrne) [inline src/jvm.jl]: I don't think you need to prefix spaces with a slash.

**Comment by aviks** [comment]
Yeah, i tried adding Compat, and got stuck on `Ref` :  https://ci.appveyor.com/project/aviks/winreg-jl/build/job/65kjixs8o02kq5jf

At this point, I would be open to dropping support for 0.3, given that 0.5 now works again. I'll probably tag a final version with other fixes before merging this PR.

**Comment by tkelman** [comment]
Be sure to use a new minor version for the first tag that drops 0.3 support

**Comment by simonbyrne** [comment]
I too tried to 0.3 support for WinReg, but got stuck as well:
https://ci.appveyor.com/project/simonbyrne/winreg-jl/build/job/d4neo01p9nm83q6e

**Comment by aviks** [comment]
Let's drop support for 0.3, and get this integrated.

**Comment by aviks** [comment]
So `Memoize` has stopped working on master again 😦 

Do we need msvcr100.dll only for winreg??

**Comment by simonbyrne** [comment]
I don't think WinReg requires any additional libraries to be loaded.

**Comment by aviks** [comment]
Spoke to @AndyGreenwell  offline. This dll seems to be required in his environment for JavaCall to work, but is not required in other environments, including appveyor. Strange!

**Comment by tkelman** [comment]
appveyor is goinf to have a lot of the runtimes already installed. andy is probably testing in a clean vm. I suspect it's java rather than anything specific to winreg that requires having a runtime library installed.

**Comment by aviks** [comment]
We are needing to explicitly `dlload`ing it prior to loading the JVM. Only on some environments. That seems dodgy. 

To clarify, this has nothing to with Winreg, I now know.

**Comment by aviks** [comment]
Well, as i said, this is really strange. On Appveyor, or another windows 7 machine i have,  it is not necessary to do this @tkelman any thoughts? Your help would be most appreciated.

**Comment by AndyGreenwell** [comment]
Have you previously installed the [Microsoft Visual C++ 2010 Redistributable Package](https://www.microsoft.com/en-us/download/details.aspx?id=14632) on those machines at some point in the past?

**Comment by tkelman** [comment]
Sure you're comparing against the exact same version of Java? I'd guess you happen to have an msvcr100 somewhere else on the path that the jvm dll is able to find, but if it's not on the path already then you need to explicitly open it.

**Comment by AndyGreenwell** [comment]
Explanation...in my new VM (and likely in my prior test environments), I did not manually set a JAVA_HOME environment variable upon installing Java.  Upon setting JAVA_HOME for the system, `using JavaCall` works without manually loading msvcr100.dll.

Two questions:

1. Would creating or modifying ENV["JAVA_HOME"] in the context of the Julia process as part of `jvm.jl` have any effect on the loading of msvcr100.dll by jvm.dll? I will try testing this possibility in the near future.

2. Should we require manual intervention by the user for setting JAVA_HOME in order to use the JavaCall.jl package?

**Comment by aviks** [comment]
this looks good. could you rebase and squash, please?

**Comment by AndyGreenwell** [comment]
Closing and will create a new PR from current master.

**Comment by AndyGreenwell** [comment]
Replaced with the following PR: https://github.com/aviks/JavaCall.jl/pull/38

**Pull request closed without merging**